File: libfgg/pool/hash_test.go
package pool

import (
	"crypto/md5"
	"fmt"
	"testing"
)

func TestHash(t *testing.T) {
	data1 := []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa")
	data2 := []byte("bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb")
	data3 := []byte("cccccccccccccccccccccccccccc")

	sum := md5.New()
	sum.Write(data1)

	fh := newFileHash()
	fh.onData(&DataChunk{
		Offset: 0,
		Length: int64(len(data1)),
	}, data1)

	if fh.sum() != fmt.Sprintf("%x", sum.Sum(nil)) {
		t.Error(fh.sum())
	}

	sum.Write(data2)

	fh.onData(&DataChunk{
		Offset: int64(len(data1)),
		Length: int64(len(data2)),
	}, data2)

	if fh.sum() != fmt.Sprintf("%x", sum.Sum(nil)) {
		t.Error(fh.sum())
	}

	// duplicate data2
	fh.onData(&DataChunk{
		Offset: int64(len(data1)),
		Length: int64(len(data2)),
	}, data2)

	if fh.sum() != fmt.Sprintf("%x", sum.Sum(nil)) {
		t.Error(fh.sum())
	}

	// duplicate data1
	fh.onData(&DataChunk{
		Offset: 0,
		Length: int64(len(data1)),
	}, data2)

	if fh.sum() != fmt.Sprintf("%x", sum.Sum(nil)) {
		t.Error(fh.sum())
	}

	sum.Write(data3)

	fh.onData(&DataChunk{
		Offset: int64(len(data1) + len(data2)),
		Length: int64(len(data3)),
	}, data3)

	if fh.sum() != fmt.Sprintf("%x", sum.Sum(nil)) {
		t.Error(fh.sum())
	}
}

File: webapp/libfgg/pool/hash.test.ts
import { assert, describe, it } from 'vitest'

import SparkMD5 from 'spark-md5'
import FileHash from "./hash"

describe('file hash test', async () => {
  const data1 = await new Blob(["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"]).arrayBuffer()
  const data2 = await new Blob(["bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"]).arrayBuffer()
  const data3 = await new Blob(["cccccccccccccccccccccccccccc"]).arrayBuffer()

  const sum = new SparkMD5.ArrayBuffer()
  sum.append(data1)

  const fh = new FileHash()
  fh.onData({
    offset: 0,
    length: data1.byteLength,
  }, data1)


  it('data1 sum', () => {
    assert.equal(fh.sum(), sum.end())
  })

  sum.append(data2)

  fh.onData({
    offset: data1.byteLength,
    length: data2.byteLength,
  }, data2)


  it('data2 sum', () => {
    assert.equal(fh.sum(), sum.end())
  })

  // duplicate data2
  fh.onData({
    offset: data1.byteLength,
    length: data2.byteLength,
  }, data2)

  it('duplicate data2 sum', () => {
    assert.equal(fh.sum(), sum.end())
  })

  // duplicate data1
  fh.onData({
    offset: 0,
    length: data1.byteLength,
  }, data2)

  it('duplicate data1 and data2 sum', () => {
    assert.equal(fh.sum(), sum.end())
  })

  sum.append(data3)

  fh.onData({
    offset: data1.byteLength + data2.byteLength,
    length: data3.byteLength,
  }, data3)

  it('data3 sum', () => {
    assert.equal(fh.sum(), sum.end())
  })
})
